site/content/entry/yuzu-progress-report-apr-2022/index.md

```diff
@@ -0,0 +1,332 @@
+	"Example in Spanish"
+	"./compatbug.png"
+	"./compatfix.png"
+>}}
+
+Similarly, D-Pad directions also didn’t translate properly. The same suspect, {{< gh-hovercard "8224" "once again" >}}.
+Someone, please send a warrant asking for the detention of Carmen Sandiego.
+
+{{< single-title-imgs
+	"Example in French"
+	"./hatbug.png"
+	"./hatfix.png"
+>}}
+
+## Kernel and CPU emulation changes
+
+Ket's begin with two changes that happened in March.
+
+Our resident bunnei rabbit continued his work on rewriting yuzu's kernel memory management to make it 
+accurate to the latest system updates. This time, he tackled and revamped {{< gh-hovercard "7974" "how the kernel code memory is mapped and unmapped" >}}.
+
+Code memory, in the context of the Switch, is where games and apps have their own code mapped for execution.
+Thanks to these changes, 'Super Smash Bros. Ultimate' no longer causes memory access issues while loading/unloading NROs, making the game stable for long runs.
+
+bunnei also {{< gh-hovercard "8013" "migrated slab heaps for the guest (Switch) kernel objects" >}} from host heap memory to emulated guest memory.
+With this change, yuzu's memory layout is now more closely matching the console.
+
+> A slab represents a contiguous piece of memory. A heap is a general term used for any memory that is
+allocated dynamically and randomly.
+
+So, slab heaps are basically pieces of memory dynamically allocated for guest kernel objects.
+By moving these away from the host (PC) heap memory (RAM) to emulated guest (Switch) memory, we can ensure
+that the kernel objects never go beyond the system limits to cause memory leaks on the host (PC).
+
+Thread local storage (TLS), the mechanism by which each thread in a given multithreaded process allocates storage for 
+thread-specific data, was also rewritten making it accurate to the latest HorizonOS behaviour.
+
+With these changes, we have now completely fixed the kernel memory object leaks that affected a few games, but went largely unnoticed, due to the previous implementation allowing unlimited allocations.
+
+Back to the list of April changes, bunnei also {{< gh-hovercard "8122" "reimplemented how yuzu handled thread allocation" >}} for HLE service interfaces.
+
+> Services are system processes running in the background which wait for incoming requests. 
+The Switch's HorizonOS has various services that perform various tasks e.g Audio, Bluetooth, etc.
+
+Previously we used to allocate one host thread per HLE service interface because -
+
+- some service routines need to block threads, and 
+- we don't support guest thread rescheduling from host threads.
+
+> A thread in block state will have to wait until an action can be completed.
+> exmaple needed
+
+The issue with this approach was that since it's the host OS that schedules these threads, yuzu could end up creating dozens of threads for services and that could lead to weird behaviour particularly on systems with hardware limitations.
```

Comment:
```suggestion
The issue with this approach was that since it's the host (Windows or Linux) that schedules the service threads, yuzu could create weird behaviour particularly on systems with hardware limitations due to the spawning one thread per service and the sheer number of service implementations we emulate.
```